File: algorithm/leetcode156.cpp
class Solution {//C++
public:
    TreeNode* upsideDownBinaryTree(TreeNode* root) {
    	if(!root || !root->left) return root;
    	TreeNode* l = root->left;//左右子节点存取来
    	TreeNode* r = root->right;
        root->left = NULL;//上下断开
        root->right = NULL;
    	TreeNode* p = upsideDownBinaryTree(l);//根节点
        l->left = r;//跟上面连接起来
        l->right = root;
        return p;
    }
};

File: algorithm/Python3_version/leetcode156.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def upsideDownBinaryTree(self, root: TreeNode) -> TreeNode:
        if not root or not root.left:
            return root
        l = root.left
        r = root.right
        root.left = None
        root.right = None
        p = self.upsideDownBinaryTree(l)
        l.left = r
        l.right = root
        return p